sleepy/types.py

```diff
@@ -436,20 +439,20 @@ class UnionType(Type):
   A tagged union, i.e. a type that can be one of a set of different types.
   """
 
-  def __init__(self, possible_types: List[Type], possible_type_nums: List[int], val_size: Optional[int],
+  def __init__(self, type_mapping: Mapping[Type, int],
+               val_size: Optional[int],
                constructor: Optional[OverloadSet] = None):
-    assert len(possible_types) == len(possible_type_nums)
-    assert len(set(possible_types)) == len(possible_types)
-    assert all(not isinstance(possible_type, UnionType) for possible_type in possible_types)
-    self.possible_types = possible_types
-    self.possible_type_nums = possible_type_nums
-    self.identifier = 'Union(%s)' % '_'.join(str(possible_type) for possible_type in possible_types)
+    assert all(not isinstance(possible_type, UnionType) for possible_type in type_mapping.values())
```

Comment:
While this wasn't here before, we should also assert sth like `assert len(set(variants_map.values())) == len(variants_map.values())` to check that the ids are unique.